lib/configuration.js

```
@@ -15,150 +15,316 @@
+      this._projectId = this._givenConfiguration.projectId;
+      cb(null, this._projectId);
+    } else if (isNumber(this._givenConfiguration.projectId)) {
+      this._projectId = this._givenConfiguration.projectId.toString();
+      cb(null, this._projectId);
+    } else {
+      cb(new Error('config.projectId must be a string or number'), null);
     }
-  } else if (!isUndefined(initConfiguration)) {
-
-    throw new Error(
-        'An invalid configuration was given to the Stackdriver Error Library');
+  } else {
+    cb(new Error([
+      'Unable to find the project Id for communication with the Stackdriver',
+      'Error Reporting service. This app will be unable to send errors to the',
+      'reporting service unless a valid project Id is supplied via runtime',
+      'configuration or the GCLOUD_PROJECT environmental variable.'
+    ].join(' ')), null);
   }
-
-  if (isString(env.GCLOUD_PROJECT) || isNumber(env.GCLOUD_PROJECT)) {
-    projectId = env.GCLOUD_PROJECT;
+};
+/**
+ * Returns the _reportUncaughtExceptions property on the instance.
+ * @memberof Configuration
+ * @public
+ * @function getReportUncaughtExceptions
+ * @returns {Boolean} - returns the _reportUncaughtExceptions property
+ */
+Configuration.prototype.getReportUncaughtExceptions = function() {
+  return this._reportUncaughtExceptions;
+};
+/**
+ * Returns the _shouldReportErrorsToAPI property on the instance.
+ * @memberof Configuration
+ * @public
+ * @function getShouldReportErrorsToAPI
+ * @returns {Boolean} - returns the _shouldReportErrorsToAPI property
+ */
+Configuration.prototype.getShouldReportErrorsToAPI = function() {
+  return this._shouldReportErrorsToAPI;
+};
+/**
+ * Returns the _projectId property on the instance.
+ * @memberof Configuration
+ * @public
+ * @function getProjectId
+ * @returns {String|Null} - returns the _projectId property
+ */
+Configuration.prototype.getProjectId = function(cb) {
+  var self = this;
+  if (!isFunction(cb)) {
```

@matthewloring @ofrobots This is an semi-philosophical question to me:
Should we maintain interfaces between our components which break quickly when used incorrectly as a way of 'defining' the interface or should they fail silently? I prefer the former since as codebases grow in size and more people begin to work on them I find that interfaces which break when used incorrectly are easier to develop with and expect certain behaviour from.